File: Practice 8/Гарифуллин/library-backend/src/categories/entities/category.entity.ts
import { Entity, ManyToOne, OneToMany, PrimaryColumn } from "typeorm";

@Entity()
export class Category {
  @PrimaryColumn({unique: true})
  name: string

  @ManyToOne(type => Category, category => category.children)
  parent: Category;

  @OneToMany(type => Category, category => category.parent)
  children: Category[];
}

File: Practice 8/Yanal Yahya/LibraryRestAPI/src/main/java/com/example/LibraryRestAPI/entity/CategoryEntity.java
package com.example.LibraryRestAPI.entity;

import javax.persistence.*;
import java.util.List;

@Entity
@Table(name = "category")
public class CategoryEntity {
    @Id
    private String categoryName;

    @OneToOne
    private CategoryEntity parentCategory;

    @OneToMany(cascade = CascadeType.ALL, mappedBy = "category")
    private List<BookCategoryEntity> readers;

    public List<BookCategoryEntity> getReaders() {
        return readers;
    }

    public void setReaders(List<BookCategoryEntity> readers) {
        this.readers = readers;
    }

    public String getCategoryName() {
        return categoryName;
    }

    public void setCategoryName(String categoryName) {
        this.categoryName = categoryName;
    }

    public CategoryEntity getParentCategory() {
        return parentCategory;
    }

    public void setParentCategory(CategoryEntity parentCategory) {
        this.parentCategory = parentCategory;
    }

    public CategoryEntity() {
    }
}
